common/iterator/FilteredIterator.java

```diff
@@ -54,4 +63,94 @@ public T next() {
     public void recycle() {
         iterator.recycle();
     }
+
+    static class Sorted<T extends Comparable<? super T>, ITER extends FunctionalIterator.Sorted<T>>
+            extends AbstractFunctionalIterator.Sorted<T> {
+
+        private final Predicate<T> predicate;
+        final ITER source;
+        T next;
+        T last;
+
+        Sorted(ITER source, Predicate<T> predicate) {
+            this.source = source;
+            this.predicate = predicate;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return (next != null) || fetchAndCheck();
+        }
+
+        private boolean fetchAndCheck() {
+            while (source.hasNext() && !predicate.test(next = source.next())) next = null;
+            return next != null;
+        }
+
+        @Override
+        public T next() {
+            if (!hasNext()) throw new NoSuchElementException();
+            last = next;
+            next = null;
+            return last;
+        }
+
+        @Override
+        public T peek() {
+            if (!hasNext()) throw new NoSuchElementException();
+            return next;
+        }
+
+        @Override
+        public void recycle() {
+            source.recycle();
+        }
+
+        static class Forwardable<T extends Comparable<? super T>>
+                extends FilteredIterator.Sorted<T, FunctionalIterator.Sorted.Forwardable<T>>
+                implements FunctionalIterator.Sorted.Forwardable<T> {
+
+            Forwardable(FunctionalIterator.Sorted.Forwardable<T> source, Predicate<T> predicate) {
+                super(source, predicate);
+            }
+
+            @Override
+            public void forward(T target) {
+                if (last != null && target.compareTo(last) < 0) throw TypeDBException.of(ILLEGAL_ARGUMENT);
+                source.forward(target);
+                next = null;
+            }
+
+            @SafeVarargs
+            @Override
+            public final FunctionalIterator.Sorted.Forwardable<T> merge(FunctionalIterator.Sorted.Forwardable<T>... iterators) {
+                return Iterators.Sorted.merge(this, iterators);
+            }
+
+            @Override
+            public <U extends Comparable<? super U>> FunctionalIterator.Sorted.Forwardable<U> mapSorted(Function<T, U> mappingFn, Function<U, T> reverseMappingFn) {
+                return Iterators.Sorted.mapSorted(this, mappingFn, reverseMappingFn);
+            }
+
+            @Override
+            public FunctionalIterator.Sorted.Forwardable<T> distinct() {
+                return Iterators.Sorted.distinct(this);
+            }
+
+            @Override
+            public FunctionalIterator.Sorted.Forwardable<T> filter(Predicate<T> predicate) {
+                return Iterators.Sorted.filter(this, predicate);
+            }
```

Comment:
we have to re-implement these in each `Forwardable` class because we can't do multiple inheritance